### ISSUE TITLE
Confusing error message with borrows and dyn Error

### ISSUE THREAD

**Issue by insanitybit** [rust-lang/rust#95404]
Given the following code: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=06baecf9bfb364dbe030e6009f2316f9
```rust
use std::sync::Arc;

async fn do_stuff(boop: Arc<dyn std::error::Error>, beep: &str, borp: &str){
}
```

The current output is:

```
error[[E0700]](https://doc.rust-lang.org/stable/error-index.html#E0700): hidden type for `impl Trait` captures lifetime that does not appear in bounds
 [--> src/lib.rs:3:76
](https://play.rust-lang.org/#)  |
3 | async fn do_stuff(boop: Arc<dyn std::error::Error>, beep: &str, borp: &str){
  |                                                                            ^
  |
  = note: hidden type `impl Future<Output = [async output]>` captures lifetime '_#14r
```

Ideally the output should look like:

I'm not really sure. I actually don't get why this is an error? If I remove any one of those arguments it works. I think this is a combination of `dyn` with async and multiple references, although weirdly if there's only one &str it compiles... idk!

Things that are confusing:

1. There is no "impl Future" - it's hidden, but where? I *know* it's because async desugares to that, but I think that's confusing.

2. `_#14r` IDK which lifetime that's referring to. I sort of suspect it's the dyn one.

3. It doesn't tell me what to do

The code to fix this is to add a lifetime:

```rust
async fn do_stuff<'a>(boop: Box<dyn std::error::Error + 'a>, beep: &str, borp: &str){
}
```

So ideally it'd suggest that. An anonymous lifetime works too, actually.

**Comment by eholk** [comment]
We talked about this in our triage meeting today. This sounds like the same issue as #63033, so I'm going to close it as a duplicate.

It looks like what's going on is that the `boop` parameter would normally be `Box<dyn std::error::Error + 'static>`, but the compiler finds an opportunity for subtyping with the future and changes it to `Box<dyn std::error::Error + 'a>`, but then it needs to pick a value for `'a`. It doesn't want to introduce a new lifetime, so instead it needs to choose between `beep` (call that `'b`) and `borp` (call that `'c`). Since `'b` and `'c` have the same constraints, there's no best choice between these two and the compiler gives up.

Fixing #63033 should fix this one too.

**Comment by aliemjay** [comment]
Thai no longer an error, fixed along with #63033.
